### PR TITLE
S3900: Add FP repro for captured parameters

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Part of #6997

This is still a draft because it documents the FP without fixing it.

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs [hunk 1]
```diff
@@ -887,6 +887,22 @@ public void Branching_IsNullOperation_Equals(string expression)
...
+            var validator = SETestContext.CreateCS("""
+                if (s == null)
+                {
+                    return;
+                }
+
+                Func<string> someFunc = () => s.ToString();
+
+                Tag("End", s);
+                """, ", string s").Validator;
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:902]: @pavel-mikula-sonarsource This is what I found out so far, but I have no idea how to fix it:

When we enter the block after the return, `s` has the NotNull constraint as expected. `s` is also in the `captured` list of `LiveVariableAnalysisBase`. Therefore it is not included in the `LiveOut` list:
```cs
        /// <summary>
        /// LiveOut variables are alive when exiting block. They are read in any of it's successors.
        /// </summary>
        public IReadOnlyCollection<ISymbol> LiveOut(TBlock block) =>
            blockLiveOut[block].Except(captured).ToImmutableArray();
```

And that is why it gets removed in `ProcessBranchState`:

https://github.com/SonarSource/sonar-dotnet/blob/0d3caf30c09f7c8e99ab783838e186dd72bf572c/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs#L165-L167

I don't understand why captured variables are not included in LiveOut and therefore I don't know how to fix it without breaking other stuff. @pavel-mikula-sonarsource, please give me directions on how to fix this.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM.
I've renamed the PR to make it explicit that it adds reproducer.

It's a draft, I'm not sure why.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:902]: This is tricky. 
LVA tracks `Captured` so we can say "We don't assume anything about them".
The old engine did not track those symbols at all.
I don't think we can properly fix it until we add proper support for lambdas and local functions. As S3900 is not a SonarWay, I would add this reproducer, accept it as FP for this sprint and create a new issue for it (if we don't have similar one).

And in the issue, capture as much information as you've gathered here.

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs:902]: New issue created and added to the backlog: #7060

**Pull request merged**